In [1]:
import numpy as np
import pandas as pd
from pydataset import data
from grab_db import my_db

## Exercises 2:

In [2]:
# Create the users DataFrame.

users = pd.DataFrame({
    'id': [1, 2, 3, 4, 5, 6],
    'name': ['bob', 'joe', 'sally', 'adam', 'jane', 'mike'],
    'role_id': [1, 2, 3, 3, np.nan, np.nan]
})
users


id   name  role_id
0   1    bob      1.0
1   2    joe      2.0
2   3  sally      3.0
3   4   adam      3.0
4   5   jane      NaN
5   6   mike      NaN

In [3]:
# Create the roles DataFrame

roles = pd.DataFrame({
    'id': [1, 2, 3, 4],
    'name': ['admin', 'author', 'reviewer', 'commenter']
})
roles


id       name
0   1      admin
1   2     author
2   3   reviewer
3   4  commenter

In [4]:
users.merge(roles, how = 'right')

id       name  role_id
0   1      admin      NaN
1   2     author      NaN
2   3   reviewer      NaN
3   4  commenter      NaN

Outer Join:

In [5]:
users.merge(roles, how = 'outer')

id       name  role_id
0   1        bob      1.0
1   2        joe      2.0
2   3      sally      3.0
3   4       adam      3.0
4   5       jane      NaN
5   6       mike      NaN
6   1      admin      NaN
7   2     author      NaN
8   3   reviewer      NaN
9   4  commenter      NaN

#### 5. Load the mpg dataset from PyDataset

In [6]:
mpg = data('mpg')


In [7]:
mpg[mpg.fl== 'd']

manufacturer               model  displ  year  cyl       trans drv  cty  \
32     chevrolet     k1500 tahoe 4wd    6.5  1999    8    auto(l4)   4   14   
123         jeep  grand cherokee 4wd    3.0  2008    6    auto(l5)   4   17   
213   volkswagen               jetta    1.9  1999    4  manual(m5)   f   33   
222   volkswagen          new beetle    1.9  1999    4  manual(m5)   f   35   
223   volkswagen          new beetle    1.9  1999    4    auto(l4)   f   29   

     hwy fl       class  
32    17  d         suv  
123   22  d         suv  
213   44  d     compact  
222   44  d  subcompact  
223   41  d  subcompact

In [8]:
mpg.fl.value_counts()

r    168
p     52
e      8
d      5
c      1
Name: fl, dtype: int64

#### 6. Output and read the documentation for the mpg dataset.

In [9]:
mpg.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 234 entries, 1 to 234
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   manufacturer  234 non-null    object 
 1   model         234 non-null    object 
 2   displ         234 non-null    float64
 3   year          234 non-null    int64  
 4   cyl           234 non-null    int64  
 5   trans         234 non-null    object 
 6   drv           234 non-null    object 
 7   cty           234 non-null    int64  
 8   hwy           234 non-null    int64  
 9   fl            234 non-null    object 
 10  class         234 non-null    object 
dtypes: float64(1), int64(4), object(6)
memory usage: 21.9+ KB


#### 7. How many rows and columns are in the dataset?

234 x 11

#### 8. Check out your column names and perform any cleanup you may want on them.

In [10]:
mpg.rename(columns = {'fl':'fuel'})

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fuel  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29    p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29    p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31    p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30    p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26    p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ...  ...   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28    p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29    p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26    p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26    p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26    p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

#### 9. Display the summary statistics for the dataset.

In [11]:
mpg.describe()

displ         year         cyl         cty         hwy
count  234.000000   234.000000  234.000000  234.000000  234.000000
mean     3.471795  2003.500000    5.888889   16.858974   23.440171
std      1.291959     4.509646    1.611534    4.255946    5.954643
min      1.600000  1999.000000    4.000000    9.000000   12.000000
25%      2.400000  1999.000000    4.000000   14.000000   18.000000
50%      3.300000  2003.500000    6.000000   17.000000   24.000000
75%      4.600000  2008.000000    8.000000   19.000000   27.000000
max      7.000000  2008.000000    8.000000   35.000000   44.000000

#### 10. How many different manufacturers are there?

In [12]:
len(mpg['manufacturer'].value_counts())

15

#### 11. How many different models are there?

In [13]:
len(mpg['model'].value_counts())

38

#### 12. Create a column named mileage_difference like you did in the DataFrames exercises; this column should contain the difference between highway and city mileage for each car.



In [14]:
mpg['mileage_difference'] = mpg['hwy'] - mpg['cty']
mpg

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  mileage_difference  
1    compact                  11  
2    compact                   8  
3    compact                  11  
4    compact                   9  
5    compact                  10  
..       ...                 ...  
230  midsize                   9  
231  midsize                   8  
232  midsize                  10  
233  midsize                   8  
234  midsize                   9  

[234 rows x 12 columns]

#### 13. Create a column named average_mileage like you did in the DataFrames exercises; this is the mean of the city and highway mileage.

In [15]:
mpg['average_mileage'] = (mpg['cty'] + mpg['hwy'])/2
mpg

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  mileage_difference  average_mileage  
1    compact                  11             23.5  
2    compact                   8             25.0  
3    compact                  11             25.5  
4    compact                   9             25.5  
5    compact                  10             21.0  
..       ...                 ...              ...  
230  midsize                   9             23.5  
231  midsize                   8             25.0  
232  midsize                  10             21.0  
233  midsize                   8             22.0  
234  midsize                   9             21.5  

[234 rows x 13 columns]

#### 14. Create a new column on the mpg dataset named is_automatic that holds boolean values denoting whether the car has an automatic transmission

In [16]:
mpg['is_auto'] = mpg.trans.str[0:4] == 'auto'
mpg

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  mileage_difference  average_mileage  is_auto  
1    compact                  11             23.5     True  
2    compact                   8             25.0    False  
3    compact                  11             25.5    False  
4    compact                   9             25.5     True  
5    compact                  10             21.0     True  
..       ...                 ...              ...      ...  
230  midsize                   9             23.5     True  
231  midsize                   8             25.0    False  
232  midsize                  10             21.0     True  
233  midsize                   8             22.0    False  
234  midsize                   9             21.5     True  

[234 rows x 14 columns]

#### 15. Using the mpg dataset, find out which which manufacturer has the best miles per gallon on average?

In [17]:
for x in np.unique(mpg.manufacturer.values):
    mpg[x] = (sum(mpg.average_mileage[mpg.manufacturer == x])/len( mpg [ mpg.manufacturer == x]))


In [18]:
for x in np.unique(mpg.manufacturer.values):
    print( mpg[x])

1      22.027778
2      22.027778
3      22.027778
4      22.027778
5      22.027778
         ...    
230    22.027778
231    22.027778
232    22.027778
233    22.027778
234    22.027778
Name: audi, Length: 234, dtype: float64
1      18.447368
2      18.447368
3      18.447368
4      18.447368
5      18.447368
         ...    
230    18.447368
231    18.447368
232    18.447368
233    18.447368
234    18.447368
Name: chevrolet, Length: 234, dtype: float64
1      15.540541
2      15.540541
3      15.540541
4      15.540541
5      15.540541
         ...    
230    15.540541
231    15.540541
232    15.540541
233    15.540541
234    15.540541
Name: dodge, Length: 234, dtype: float64
1      16.68
2      16.68
3      16.68
4      16.68
5      16.68
       ...  
230    16.68
231    16.68
232    16.68
233    16.68
234    16.68
Name: ford, Length: 234, dtype: float64
1      28.5
2      28.5
3      28.5
4      28.5
5      28.5
       ... 
230    28.5
231    28.5
232    28.5
233    28.5
234    28.

HONDA

#### 16. Do automatic or manual cars have better miles per gallon?

In [19]:
auto = (sum( mpg.average_mileage[mpg.is_auto == True]/ len(mpg[mpg.is_auto == True])))
auto

19.130573248407636

In [20]:
manual = (sum(mpg.average_mileage[mpg.is_auto == False]/len(mpg[mpg.is_auto == False])))
manual

22.227272727272734

Manuals have better average mileage

# Part 3:

#### 1. Use your get_db_url function to help you explore the data from the chipotle database.

In [2]:
chipotle = my_db('''Select * From orders''', 'chipotle')
chipotle

id  order_id  quantity                              item_name  \
0        1         1         1           Chips and Fresh Tomato Salsa   
1        2         1         1                                   Izze   
2        3         1         1                       Nantucket Nectar   
3        4         1         1  Chips and Tomatillo-Green Chili Salsa   
4        5         2         2                           Chicken Bowl   
...    ...       ...       ...                                    ...   
4617  4618      1833         1                          Steak Burrito   
4618  4619      1833         1                          Steak Burrito   
4619  4620      1834         1                     Chicken Salad Bowl   
4620  4621      1834         1                     Chicken Salad Bowl   
4621  4622      1834         1                     Chicken Salad Bowl   

                                                                     choice_description  \
0                                                                                   nan   
1                                                                          [Clementine]   
2                                                                               [Apple]   
3                                                                                   nan   
4            [Tomatillo-Red Chili Salsa (Hot), [Black Beans, Rice, Cheese, Sour Cream]]   
...                                                                                 ...   
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour Cream, Cheese, Lettuce, Guacamole]]   
4618               [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese, Lettuce, Guacamole]]   
4619         [Fresh Tomato Salsa, [Fajita Vegetables, Pinto Beans, Guacamole, Lettuce]]   
4620                                 [Fresh Tomato Salsa, [Fajita Vegetables, Lettuce]]   
4621                    [Fresh Tomato Salsa, [Fajita Vegetables, Pinto Beans, Lettuce]]   

     item_price  
0        $2.39   
1        $3.39   
2        $3.39   
3        $2.39   
4       $16.98   
...         ...  
4617    $11.75   
4618    $11.75   
4619    $11.25   
4620     $8.75   
4621     $8.75   

[4622 rows x 6 columns]

#### 2. What is the total price for each order?

In [22]:
chipotle['item_price'] = chipotle['item_price'].str.strip('$')

In [23]:
chipotle['item_price'] = pd.to_numeric(chipotle.item_price)

In [24]:
chipotle['order_total'] = chipotle.order_id
for x in np.unique(chipotle.order_id.values):
    chipotle.order_total[chipotle.order_id == x] = sum(chipotle.item_price[chipotle.order_id == x])
chipotle

/var/folders/yn/3nqd95h90wd7cdqr1w2mwmz40000gn/T/ipykernel_38031/2372362890.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  chipotle.order_total[chipotle.order_id == x] = sum(chipotle.item_price[chipotle.order_id == x])


Unnamed: 0    id  order_id  quantity  \
0              0     1         1         1   
1              1     2         1         1   
2              2     3         1         1   
3              3     4         1         1   
4              4     5         2         2   
...          ...   ...       ...       ...   
4617        4617  4618      1833         1   
4618        4618  4619      1833         1   
4619        4619  4620      1834         1   
4620        4620  4621      1834         1   
4621        4621  4622      1834         1   

                                  item_name  \
0              Chips and Fresh Tomato Salsa   
1                                      Izze   
2                          Nantucket Nectar   
3     Chips and Tomatillo-Green Chili Salsa   
4                              Chicken Bowl   
...                                     ...   
4617                          Steak Burrito   
4618                          Steak Burrito   
4619                     Chicken Salad Bowl   
4620                     Chicken Salad Bowl   
4621                     Chicken Salad Bowl   

                                                                     choice_description  \
0                                                                                   NaN   
1                                                                          [Clementine]   
2                                                                               [Apple]   
3                                                                                   NaN   
4            [Tomatillo-Red Chili Salsa (Hot), [Black Beans, Rice, Cheese, Sour Cream]]   
...                                                                                 ...   
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour Cream, Cheese, Lettuce, Guacamole]]   
4618               [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese, Lettuce, Guacamole]]   
4619         [Fresh Tomato Salsa, [Fajita Vegetables, Pinto Beans, Guacamole, Lettuce]]   
4620                                 [Fresh Tomato Salsa, [Fajita Vegetables, Lettuce]]   
4621                    [Fresh Tomato Salsa, [Fajita Vegetables, Pinto Beans, Lettuce]]   

      item_price  order_total  
0           2.39        11.56  
1           3.39        11.56  
2           3.39        11.56  
3           2.39        11.56  
4          16.98        16.98  
...          ...          ...  
4617       11.75        23.50  
4618       11.75        23.50  
4619       11.25        28.75  
4620        8.75        28.75  
4621        8.75        28.75  

[4622 rows x 8 columns]

#### 3. What are the most popular 3 items?

In [38]:
top_three = chipotle.groupby('item_name').quantity.sum().sort_values(ascending=False).head(3)
top_three

item_name
Chicken Bowl           761
Chicken Burrito        591
Chips and Guacamole    506
Name: quantity, dtype: int64

#### 4. Which item has produced the most revenue?

In [41]:
(chipotle.groupby('item_name').quantity.sum().sort_values() * chipotle.groupby('item_name').item_price.sum()).sort_values(ascending = False).head()

item_name
Chicken Bowl           5587817.53
Chicken Burrito        3295309.62
Steak Burrito          1486651.98
Chips and Guacamole    1113726.24
Steak Bowl              499501.99
dtype: float64

#### 5. Join the employees and titles DataFrames together.

In [3]:
employees = my_db('Select * from employees','employees')
titles = my_db('Select * from titles', 'employees')

In [4]:
employees

emp_no  birth_date first_name last_name gender   hire_date
0        10001  1953-09-02     Georgi   Facello      M  1986-06-26
1        10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2        10003  1959-12-03      Parto   Bamford      M  1986-08-28
3        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4        10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12
...        ...         ...        ...       ...    ...         ...
300019  499995  1958-09-24     Dekang  Lichtner      F  1993-01-12
300020  499996  1953-03-07       Zito      Baaz      M  1990-09-27
300021  499997  1961-08-03    Berhard    Lenart      M  1986-04-21
300022  499998  1956-09-05   Patricia   Breugel      M  1993-10-13
300023  499999  1958-05-01     Sachin   Tsukuda      M  1997-11-30

[300024 rows x 6 columns]

In [5]:
titles

emp_no            title   from_date     to_date
0        10001  Senior Engineer  1986-06-26  9999-01-01
1        10002            Staff  1996-08-03  9999-01-01
2        10003  Senior Engineer  1995-12-03  9999-01-01
3        10004         Engineer  1986-12-01  1995-12-01
4        10004  Senior Engineer  1995-12-01  9999-01-01
...        ...              ...         ...         ...
443303  499997         Engineer  1987-08-30  1992-08-29
443304  499997  Senior Engineer  1992-08-29  9999-01-01
443305  499998     Senior Staff  1998-12-27  9999-01-01
443306  499998            Staff  1993-12-27  1998-12-27
443307  499999         Engineer  1997-11-30  9999-01-01

[443308 rows x 4 columns]

In [9]:
employees = employees.merge(titles, left_on= 'emp_no', right_on= 'emp_no', how = 'inner')

#### 6. For each title, find the hire date of the employee that was hired most recently with that title.

In [23]:
for x in np.unique(employees.title.values):
    print(employees.hire_date[(employees.title == x)&(employees.hire_date == max(employees.hire_date[employees.title == x]))], x)

212605    1999-12-24
Name: hire_date, dtype: object Assistant Engineer
389997    2000-01-28
Name: hire_date, dtype: object Engineer
148155    1992-02-05
Name: hire_date, dtype: object Manager
145474    2000-01-01
Name: hire_date, dtype: object Senior Engineer
182100    2000-01-13
Name: hire_date, dtype: object Senior Staff
55137    2000-01-12
Name: hire_date, dtype: object Staff
287947    1999-12-31
Name: hire_date, dtype: object Technique Leader


#### 7. 